pkg/api/v1/clients/multicluster/client_manager.go

```diff
@@ -0,0 +1,92 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/solo-io/go-utils/contextutils"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients"
+	"github.com/solo-io/solo-kit/pkg/multicluster/handler"
+	"go.uber.org/zap"
+	"k8s.io/client-go/rest"
+)
+
+//go:generate mockgen -destination=./mocks/client_manager.go -source client_manager.go -package mocks
+
+// Allows objects to register callbacks with a ClusterClientManager.
+type ClusterClientHandler interface {
+	HandleNewClusterClient(cluster string, client clients.ResourceClient)
+	HandleRemovedClusterClient(cluster string, client clients.ResourceClient)
+}
+
+// Stores clients for clusters as they are discovered by a config watcher.
+// Implementation in this package allows registration of ClusterClientHandlers.
+type ClusterClientManager interface {
+	handler.ClusterHandler
+	// Returns a client for the given cluster if one exists.
+	ClientForCluster(cluster string) (client clients.ResourceClient, found bool)
+}
+
+type clusterManager struct {
+	ctx            context.Context
+	clientGetter   ClientGetter
+	clientHandlers []ClusterClientHandler
+	clients        map[string]clients.ResourceClient
+	clientAccess   sync.RWMutex
+}
+
+func NewClusterClientManager(ctx context.Context, clientGetter ClientGetter, handlers ...ClusterClientHandler) ClusterClientManager {
```

Comment:
nit: return struct pointer rather than interface (https://mycodesmells.com/post/accept-interfaces-return-struct-in-go)

Reply:
will do, but I think we miss out on some of the benefits of this pattern by not also exporting the struct.